### ISSUE TITLE
7.08 is incompatible with coq 8.9.1 (and others?)

### ISSUE THREAD

**Issue by garrigue** [camlp5/camlp5#45]
Trying to install coq with opam in a 4.08.0 switch gives the following error message:
```
opam install -v coq
...
- CAMLP5O   plugins/ssrmatching/g_ssrmatching.ml4
- Error while loading “/Users/garrigue/.opam/4.08.0/lib/camlp5/pr_dump.cmo”: The module `Pr_dump’ is already loaded (either by the main program or a previously-dynlinked library).
```
Then in my case it hangs on computing dependencies.
Commenting out
```
PR_O := $(if $(READABLE_ML4),pr_o.cmo,pr_dump.cmo)
```
in `Makefile.build` solves the problem.
If there was no change in camlp5, this may mean that 4.08.0 is stricter about dynamic loading.
I'm not sure whether `pr_dump.cmo` has to be in `camlp5o` or not, but it might be better to remove it in versions after 4.08.0 to avoid this problem.

**Comment by roglo** [comment]
pr_dump.cmo may not be in camlp5o but is still part of camlp5 distribution.

**Comment by garrigue** [comment]
I meant that apparently in 7.08 pr_dump.cmo is statically  linked into camkp5o, and that causes a dynamic linking error when compiling coq.

**Comment by roglo** [comment]
I made no change about that.

**Comment by garrigue** [comment]
This may come from a change in ocaml 4.08.0.
Apparently, 4.08.1 contains a fix for dynamic loading of packed modules, but this seems irrelevant here.

**Comment by glondu** [comment]
See also https://github.com/coq/coq/issues/10403

**Comment by roglo** [comment]
Ok, if someone has a patch for that for Camlp5, make a pull request.

**Comment by garrigue** [comment]
I'm starting to think that the problem is in ocaml or coq rather than camlp5.
Correct me if I'm wrong, but `camlp5o` already contains `pr_dump.cmo`, but not `pr_o.cmo`.
So the line
```
PR_O := $(if $(READABLE_ML4),pr_o.cmo,pr_dump.cmo)
```
in coq 8.9.1 `Makefile.build` is wrong, and should actually be
```
PR_O := $(if $(READABLE_ML4),pr_o.cmo,)
```

This said, there may be other software that depended on the more lax older behavior in ocaml.
(That is, the only explanation I see is that ocaml 4.07 didn't fail when loading an already present module)
This is all hypothesis, as I didn't have time to test that yet.

**Comment by garrigue** [comment]
I've proposed PRs for coq v8.9 (https://github.com/coq/coq/pull/10650) and directly through opam (https://github.com/ocaml/opam-repository/pull/14668)
Unfortunately, opam has already moved to 4.08.1, so the opam CI will only test that once camlp5 is updated to 4.08.1.
By the way, it may be that 4.08.1 fixes the problem already, but I see nothing relevant in the Changes files (it fixes dynamic loading, but only for packed modules).

**Comment by garrigue** [comment]
I opened an issue on ocaml too: https://github.com/ocaml/ocaml/issues/8868

**Comment by garrigue** [comment]
This seems to be a consequence of https://github.com/ocaml/ocaml/pull/1063, and since it was a soundness fix, this probably won't be rolled back...

**Comment by garrigue** [comment]
I added PR #46 that fixes the problem, at least for coq 8.9.1, and potentially for other users too. 
It might be a good idea to include it in the 4.08.1 compatible release of camlp5.

**Comment by XVilka** [comment]
Was this solved then with the release of Coq-8.10 and Camlp5-7.10?

**Comment by garrigue** [comment]
Yes, this is solved in 7.10.
